src/track/beatgrid.cpp

```diff
@@ -184,13 +192,56 @@ double BeatGrid::findNthBeat(double dSamples, int n) const {
         n = n + 1;
     }
 
-    double dResult = dClosestBeat + n * m_dBeatLength;
+    double dResult = floor(dClosestBeat + n * m_dBeatLength);
     if (!even(static_cast<int>(dResult))) {
         dResult--;
     }
     return dResult;
 }
 
+bool BeatGrid::findPrevNextBeats(double dSamples,
+                                 double* dpPrevBeatSamples,
+                                 double* dpNextBeatSamples) const {
+    double dFirstBeatSample;
+    {
+        QMutexLocker locker(&m_mutex);
+        if (!isValid()) {
+            *dpPrevBeatSamples = -1.0;
+            *dpNextBeatSamples = -1.0;
+            return false;
+        }
+        dFirstBeatSample = firstBeatSample();
+    }
+
+    double beatFraction = (dSamples - dFirstBeatSample) / m_dBeatLength;
```

Comment:
this (and later code) is touching m_dBeatLength outside the critical section  -- you could capture it from within the locked scope as a local variable and then use it later